### PR TITLE
Add wearable body cameras

### PR DESCRIPTION
**Pull request by Rainfey**
## About the PR
Added body wearable cameras

## Why / Balance
Seemed like a good idea, allows command to monitor individual crew or to monitor in progress operations without putting themselves in danger
Also - crew entertainment

## Technical details
Cameras are equippable to the neck slot.
The cameras broadcast on the Wireless device net, on the Security subnet.
This necessitates as Wireless Camera Router [Security] be mapped OR built by the crew
Each body camera otherwise operates the same as a news camera.
Cameras are automatically activated when equipped, and deactivated when unequipped ( i feel small spy cameras to be a bit unbalanced, so they only work when equipped)
They can be constructed at the Security Techfab for 1 Steel, 1 Glass and 2 Plastic.

Alternatives to the above aproach have their own drawbacks:
1) Bodycams could be added to the Wired device net, this will give them unlimited range and prevent audio (Camera Monitor does not output audio from what i can see)
2) Bodycams could be added to a new device net, which will require a new computer/device to receive and display the feed.

## Media
Camera when worn
Screenshot 2023-11-26 214138

Camera on the ground (crowbar for scale)
Screenshot 2023-11-26 214025

Video of usage

https://github.com/space-wizards/space-station-14/assets/11758391/b1aea924-e569-41ca-8cf2-3d05fcca4fc4

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**
:cl:
- add: Wearable Bodycameras

### CODE REVIEW

**Comment by Brandon-Huu** [comment]
Now we need the ability to record and replay

**Comment by lzk228** [comment]
i was trying to add some previously in #17324 
but the issue is cameras don't use energy, what looks kinda OP

**Comment by UbaserB** [comment]
I feel like the sprite should be improved, right now it looks like it doesnt fit the style of the game for some reason

**Comment by Rainfey** [comment]
I make no claims of being any good with artwork - consider them placeholders.

**Comment by UbaserB** [comment]
> I make no claims of being any good with artwork - consider them placeholders. Regarding battery power, this seems reasonable - if its deemed necessary then i shall add it in.

i mean if you want i can try sprite it

**Comment by Ilya246** [comment]
should have an implant version

**Comment by Hmeister-fake** [comment]
If this is added, first off it is way too high quality to be fair, why do reporters need to lug around big cameras when security has the same thing. It would be better if the cameras where intentionally delayed and had some grainy filter, maybe also don't transmit sound just video. Also they should probably be more zoomed in, so that you can't see 360 around the person wearing it. Other than that, cool PR.

**Comment by IamVelcroboy** [comment]
> If this is added, first off it is way too high quality to be fair, why do reporters need to lug around big cameras when security has the same thing. It would be better if the cameras where intentionally delayed and had some grainy filter, maybe also don't transmit sound just video. Also they should probably be more zoomed in, so that you can't see 360 around the person wearing it. Other than that, cool PR.

Or just make reporter's cameras better.

**Comment by Rainfey** [comment]
I dont think there is a way to reduce the visible FOV on a camera, its always 360.
But i think i can zoom it in.
I can also reduce the wireless range, currently they work within a range of 200 (tiles?) from the camera router - meaning there would be deadzones.
Same goes for the microphone, can reduce the range on that.

**Comment by Rainfey** [comment]
Added battery slot and usage, calibrated for 10 minutes per small cell
Added popup text when activating/deactivating the camera
Camera can be toggled when worn, without holding in your hand
Reduced audio pickup range to 5 rather than 10
Fixed license on texture

Question: Do i need an 'ActiveBodyCameraComponent' like the radio jammer? 
Not sure if there was a reason a new component was used versus an Enabled bool.

**Comment by dvir001** [comment]
https://github.com/new-frontiers-14/frontier-station-14/pull/312

Seem about like what I did, only thing you might want to change is require it to be equipped to work as people will just hide them around if not.

**Comment by Rainfey** [comment]
Had a look over your PR dvir001, some nice ideas.

- Cameras are auto enabled when equipped, and disabled when unequipped - preventing them from working in bags or pockets.
(Also means you cant leave the camera on the floor as a spy camera, perhaps this is a good idea but i think it would be a little overpowered to throw cameras into cargo 'just in case' there are Rev's)
- Cameras now set their name based upon your ID, had to touch SurveillanceCameraSystem for this - added a public method for changing names.
- Removed the UI from the camera, since names are auto set.

File: Content.Server/Bodycamera/BodyCameraSystem.cs
Comment (metalgearsloth) [inline Content.Server/Bodycamera/BodyCameraSystem.cs:102]: PowerCellDraw has useage code for this.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
1. You need to move your code to shared so it's predicted.
2. Your code needs more documenting to explain stuff, e.g. BodyCamera isn't even documented.
3. Please use file-scoped namespaces.

Any maintainers that review this bear in mind it's adding PVS viewersubs which are extremely expensive.

**Comment by Rainfey** [comment]
Im not sure if i can move BodyCameraSystem into Content.Shared, it uses SurveillanceCameraSystem which only exists in Content.Server.
SharedSurveillanceCameraSystem.cs does exist, but it does not appear to contain any classes - just two enums.

Documentation first pass, is more detail or additional documentation required?
Now using PowerCellDraw.
Now using file-scoped namespaces

File: Content.Server/Bodycamera/BodyCameraSystem.cs
Comment (Rainfey) [reply Content.Server/Bodycamera/BodyCameraSystem.cs]: Corrected, now using PowerCellDraw.

**Comment by Skarletto** [comment]
Half a year ago I made bodycam sprites for one of Chief_Engineer's prototype uploads if this may interest you, otherwise (YES!!) finally bodycams (epic)

equipped-NECK
icon2

**Comment by Skarletto** [comment]
Just standard CC-BY-SA 3.0

**Comment by ps3moira** [comment]
I would suggest making the camera's view range a lot smaller. Bit weird it has such a large range being so small.

**Comment by Rainfey** [comment]
Dont know if it ever will, it seems it may cause performance issues - or atleast contribute to them.
I'll keep it maintained and mergeable for now.

**Comment by metalgearsloth** [comment]
> I'll keep it maintained and mergeable for now.

You don't need to resolve merge conflicts until after a PR is approved as it'll just waste time.

File: Content.Client/Bodycamera/BodyCameraSystem.cs
Comment (metalgearsloth) [inline Content.Client/Bodycamera/BodyCameraSystem.cs:40]: This systems looks like it shouldn't exist realistically.
1. Dummy initialize method.
2. All it's doing is playing predicted audio.

File: Content.Server/Bodycamera/BodyCameraSystem.cs
Comment (metalgearsloth) [inline Content.Server/Bodycamera/BodyCameraSystem.cs:39]: Use TryComp and pass it in because resolving it twice is just going to be twice as expensive.

File: Content.Server/Bodycamera/BodyCameraSystem.cs
Comment (metalgearsloth) [inline Content.Server/Bodycamera/BodyCameraSystem.cs:79]: This should be using PowerCellChangedEvent looks like and should also be able to go to shared.

File: Content.Shared/Bodycamera/BodyCameraComponent.cs
Comment (metalgearsloth) [inline Content.Shared/Bodycamera/BodyCameraComponent.cs:32]: Needs networking + datafields, also the distances should really be a constant on contentaudiosystem so we have at least some consistency between systems.

File: Content.Server/Bodycamera/BodyCameraSystem.cs
Comment (metalgearsloth) [inline Content.Server/Bodycamera/BodyCameraSystem.cs:87]: Can't you just check if it's equipped here instead of manually tracking a bool which is prone to bugs?

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Comment by AJCM-git** [comment]
@Rainfey are you coming back to this?

**Comment by Rainfey** [comment]
Indeed i am, once time permits.
Work priorities have cut into my available free time.

File: Content.Server/Bodycamera/BodyCameraSystem.cs
Comment (Rainfey) [reply Content.Server/Bodycamera/BodyCameraSystem.cs]: Found a problem with PowerCellChangedEvent and HasDrawCharge.
The field HasDrawCharge relies upon has not been sync'd when PowerCellChanged is raised after inserting a battery
So it always returns false.
Worked around it by disallowing changing the battery while the camera is equipped.

File: Content.Shared/Bodycamera/BodyCameraComponent.cs
Comment (Rainfey) [reply Content.Shared/Bodycamera/BodyCameraComponent.cs]: What those constant values are will probably need some discussion, along with what naming scheme should be used (ie Quiet, Normal, Loud? or just values like 2, 5, 10)
Either way i feel its a bit out of scope for this PR.

**Review by AJCM-git — CHANGES_REQUESTED** [review]
So, taking a quick glance, the body camera system seems very unnecesary, as far as i see it serves these purposes
1) Playing sounds/popups on power on/off
2) Showing a examined message to anyone wearing one
3) Activating/Deactivating depending on power
4) Enabling the camera when worn
etc

All in all its a pretty singleton component for no good reason, split it up or add functionalities to current systems/components. For example:
- The current camera system should prob handle the batteries, no reason to split that
- The popup/sound when powered on/off should be added to the PowerCellSlotComponent, one of them also plays when wearing the item i think, for that ClothingComponent already has a sound datafield IIRC, 
- The examine should prob be a field in the PowerCellSlotComponent
- If a functionality doesnt fit a current component just make a new one handling that specifically, with an appropiatte name (if you cant figure out how to split something just leave it in for now until next review)

Also this is way too busted for it to be printable, the fact that heads have to go by themselves to check on people is intended and an important part of the game, just add 2 or 3 to the HoS locker or something (tho if you can think of something better thats cool too, i also had the idea of making them broadcast to a public channel so the down side is everyone knowing your location at all times, but heads and sec could just turn them off/on depending on whether or not they are in danger which defeats the point). Also related to this, this is so important for antags to know that examining someone should tell you if someone has a bodycam in them and if its on or off (do confirm if EMPs disable em too)

**Comment by Rainfey** [comment]
Drafting as the suggested changes are extensive.

**Comment by Emisse** [comment]
Closed due to feature freeze May 10th-June 14th. Comment to have it reopen after this.

**Comment by ps3moira** [comment]
Just realized this would be awesome if there was an alternative sprite for a hand-held live camera for news reporters!

**Comment by GroggleG** [comment]
This would be an actual fucking godsend for upcoming AI, keep this alive.

**Pull request closed without merging**
